### ISSUE TITLE
Question: Object Manager objects' get methods return 1

### ISSUE THREAD

**Issue by CoreyCole** [magento/magento2#1760]
Hello, I'm trying to print data to the success page about the most recent order placed. In `app/code/Magento/Checkout/Controller/Onepage/Success.php` I've added the following php:
```php
<?php
$orderID = $session->getLastOrderId();
$orderData = $this->_objectManager->create('Magento\Sales\Model\Order')->loadByIncrementId($orderID);
$orderItems = $orderData->getAllItems();
$orderString = "";
foreach ($orderItems as $item) {
    $orderString = $orderString.$item->getSku();
}
$orderAddress = $orderData->getShippingAddress();
?>
<div id="order-data">
    <pre id="order-number"><?= $orderID ?></pre>
    <pre id="order-items"><?= print_r($orderItems) ?></pre>
    <pre id="order-items-string"><?= print_r($orderString) ?></pre>
    <pre id="order-address"><?= print_r($orderAddress) ?></pre>
    <pre id="order-name"><?= print_r($orderData->getFirstName()) ?></pre>
</div>
```
I know it's working to some degree. The page is loading without errors and in the admin panel, I confirmed that the order is being placed with the printed $orderID with valid data. However, only the $orderID seems to be printing correctly. The output div comes out as follows:
```html
<div id="order-data">
    <!-- correct ID -->
    <pre id="order-number">63</pre> 
    <!-- looks like an empty array followed by a 1? -->
    <pre id="order-items">
        Array
        (
        )
        1
    </pre> 
    <pre id="order-items-string">1</pre> <!-- all the other get methods return "1" -->
    <pre id="order-address">1</pre>
    <pre id="order-name">1</pre>
</div>
```
Sorry if this is a rudimentary question, but I've been struggling with this for a while. I can't find any pages in the php developer guide docs about interacting with the model. What am I doing wrong?

**Comment by Vinai** [comment]
At least one thing that is wrong on first glance is that you are printing the return value of `[print_r()](http://php.net/manual/en/function.print-r.php)`, which by default is a boolean `true`.
In order to make `print_r()` return the string representation of the argument, you have to pass a truthy second argument, e.g. `<?= print_r($orderItems, true) ?>`.

**Comment by CoreyCole** [comment]
Thank @Vinai! I was not aware of the second param to print_r. I figured out my error though.
```php
$orderID = $session->getLastOrderId()
```
returns an integer while
```php
loadByIncrementId()
```
looks for the string with leading zeros. Calling:
```php
loadByIncrementId("0000000".$orderID);
```
temporarily fixed it, but I'll look for a method that returns the id as the full string.

**Comment by Vinai** [comment]
Great! Just a plain old `$order->load($orderId)` should do the trick.
By the way, are you aware that using the object manager directly to instantiate the objects you want is not the right way to do things?

**Comment by CoreyCole** [comment]
No I was unaware, what is the better way?

**Comment by Vinai** [comment]
You simply declare the dependencies you need in your class constructor, and they will be automatically injected when your class in created by the object manager.
Entities like orders however are a special case, so called non-injectables. Even though technically you could have them injected just like any other object, you want to inject a entity factory in such a case.
For example, if you want access to an order, you would declare a dependency on the `OrderFactory` in your constructor.

```php
public function __construct(\Magento\Sales\Model\OrderFactory $orderFactory)
{
    $this->orderFactory = $orderFactory;
}
```
Now, every time you need an order instance, you call 
```php
$order = $this->orderFactory->create()->load($orderId);
```

Since you where working in a core class and outputting HTML in a PHP class I assumed you where just playing around, which is fine. But if you want to do real customizations, that is not the best thing to do.

**Comment by CoreyCole** [comment]
Okay Thanks! I'm trying that, but having trouble. Is there something else I also need to do? Since I'm using this on the Order Success page I added that dependency in `/app/code/Magento/Checkout/Controller/Onepage.php`, but I'm getting the following error:

`Recoverable Error: Argument 15 passed to Magento\Checkout\Controller\Onepage::__construct() must be an instance of Magento\Sales\Model\OrderFactory, none given, called in /var/www/html/magento2/var/generation/Magento/Checkout/Controller/Index/Index/Interceptor.php on line 14 and defined in /var/www/html/magento2/app/code/Magento/Checkout/Controller/Onepage.php on line 118`

I tried clearing out my `var` folder because I saw `Interceptor.php` there, but that didn't help.

**Comment by CoreyCole** [comment]
Thanks for all the help @Vinai! That worked, but for anyone else looking at this, I also had to add the field above the constructor.
```php
/**
* @var \Magento\Sales\Model\OrderFactory
*/
protected $orderFactory;
```

**Comment by magento-engcom-team** [comment]
Hi @CoreyCole. Thank you for your report. 
The issue has been fixed in magento/magento2#29632 by @jmonteros422 in 2.4-develop branch
Related commit(s):
 - [afe499d5fbd09abd86e3c8d48797395fada01fa6](https://github.com/magento/magento2/commit/afe499d5fbd09abd86e3c8d48797395fada01fa6)
 - [0994c44a964406f1eae1c4779903c78533111989](https://github.com/magento/magento2/commit/0994c44a964406f1eae1c4779903c78533111989)
 - [5a70b81a23733856bfd7d4e74e23106f97d245d9](https://github.com/magento/magento2/commit/5a70b81a23733856bfd7d4e74e23106f97d245d9)
 - [22670212eee321623260da8e9b406c8186309ed3](https://github.com/magento/magento2/commit/22670212eee321623260da8e9b406c8186309ed3)

The fix will be available with the upcoming 2.4.2 release.
